<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>seqio API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>seqio</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import gzip
import random
import sys

def read_fasta(filename):
        &#34;&#34;&#34;
        *Function that reads fasta files* &lt;br/&gt;

        *Returns sequence name followed by sequences from the imported file* &lt;br/&gt;

        **Parameters:**
        _______________

        + filename -- path to the file (str)
        &#34;&#34;&#34;

        name = None
        seqs = []

        fp = None
        if filename == &#39;-&#39;:
                fp = sys.stdin
        elif filename.endswith(&#39;.gz&#39;):
                fp = gzip.open(filename, &#39;rt&#39;)
        else:
                fp = open(filename)

        for line in fp.readlines():
                line = line.rstrip()
                if line.startswith(&#39;&gt;&#39;):
                        if len(seqs) &gt; 0:
                                seq = &#39;&#39;.join(seqs)
                                yield(name, seq)
                                name = line[1:]
                                seqs = []
                        else:
                                name = line[1:]
                else:
                        seqs.append(line)
        yield(name, &#39;&#39;.join(seqs))
        fp.close()

def fasta2onehot(file, label):
        &#34;&#34;&#34;
        *Function converts sequences stored in fasta format into one-hot encoded data* &lt;br/&gt;

        *Returns one-hot encoded sequences* &lt;br/&gt;

        **Parameter:**
        ______________

        + file -- path to the fasta file containing sequences (str)
        + label -- label provided by the use (int)
        &#34;&#34;&#34;

        data = []
        for name, seq in read_fasta(file):
                s = &#39;&#39;
                for nt in seq:
                        if   nt == &#39;A&#39;: s += &#39;1000&#39;
                        elif nt == &#39;C&#39;: s += &#39;0100&#39;
                        elif nt == &#39;G&#39;: s += &#39;0010&#39;
                        elif nt == &#39;T&#39;: s += &#39;0001&#39;
                        else: raise()
                s += str(label)
                data.append(s)
        return data

def read_raw(filename):
        &#34;&#34;&#34;
        *Function that returns sequencing data in fasta format* &lt;br/&gt;

        *Returns unfiltered names and sequences* &lt;br/&gt;

        **Parameter:**
        ______________

        + filename -- path to the fasta file containing sequences (str)
        &#34;&#34;&#34;

        fp = None
        if filename == &#39;-&#39;:
                fp = sys.stdin
        elif filename.endswith(&#39;.gz&#39;):
                fp = gzip.open(filename, &#39;rt&#39;)
        else:
                fp = open(filename)

        for line in fp.readlines():
                line = line.rstrip()
                yield(line)

        fp.close()

def random_dna(length):
        &#34;&#34;&#34;
        *Function that generates random dna sequences based on desired length (not weighted)* &lt;br/&gt;

        *Returns random dna sequence based on input length (str)* &lt;br/&gt;

        **Parameter:**
        ______________

        + length -- length of generated dna sequence provided by the user (int)

        &#34;&#34;&#34;
        nts = &#39;ACGT&#39;
        seq = []
        for i in range(length):
                seq.append(random.choice(nts))
        return &#39;&#39;.join(seq)

def cross_validation(seqs, x):
        &#34;&#34;&#34;
        *Function that generated train and test set of a provided dataset* &lt;br/&gt;

        *Returns train and test sets (list), (list)* &lt;br/&gt;

        **Parameter:**
        ______________
        + seqs -- list of sequences (list)
        + x -- number of cross validations (int)
        &#34;&#34;&#34;
        for i in range(x):
                train = []
                test = []
                for j in range(len(seqs)):
                        if j % x == i: test.append(seqs[j])
                        else:          train.append(seqs[j])
                yield train, test</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="seqio.cross_validation"><code class="name flex">
<span>def <span class="ident">cross_validation</span></span>(<span>seqs, x)</span>
</code></dt>
<dd>
<div class="desc"><p><em>Function that generated train and test set of a provided dataset</em> <br/></p>
<p><em>Returns train and test sets (list), (list)</em> <br/></p>
<p><strong>Parameter:</strong></p>
<hr>
<ul>
<li>seqs &ndash; list of sequences (list)</li>
<li>x &ndash; number of cross validations (int)</li>
</ul></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cross_validation(seqs, x):
        &#34;&#34;&#34;
        *Function that generated train and test set of a provided dataset* &lt;br/&gt;

        *Returns train and test sets (list), (list)* &lt;br/&gt;

        **Parameter:**
        ______________
        + seqs -- list of sequences (list)
        + x -- number of cross validations (int)
        &#34;&#34;&#34;
        for i in range(x):
                train = []
                test = []
                for j in range(len(seqs)):
                        if j % x == i: test.append(seqs[j])
                        else:          train.append(seqs[j])
                yield train, test</code></pre>
</details>
</dd>
<dt id="seqio.fasta2onehot"><code class="name flex">
<span>def <span class="ident">fasta2onehot</span></span>(<span>file, label)</span>
</code></dt>
<dd>
<div class="desc"><p><em>Function converts sequences stored in fasta format into one-hot encoded data</em> <br/></p>
<p><em>Returns one-hot encoded sequences</em> <br/></p>
<p><strong>Parameter:</strong></p>
<hr>
<ul>
<li>file &ndash; path to the fasta file containing sequences (str)</li>
<li>label &ndash; label provided by the use (int)</li>
</ul></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def fasta2onehot(file, label):
        &#34;&#34;&#34;
        *Function converts sequences stored in fasta format into one-hot encoded data* &lt;br/&gt;

        *Returns one-hot encoded sequences* &lt;br/&gt;

        **Parameter:**
        ______________

        + file -- path to the fasta file containing sequences (str)
        + label -- label provided by the use (int)
        &#34;&#34;&#34;

        data = []
        for name, seq in read_fasta(file):
                s = &#39;&#39;
                for nt in seq:
                        if   nt == &#39;A&#39;: s += &#39;1000&#39;
                        elif nt == &#39;C&#39;: s += &#39;0100&#39;
                        elif nt == &#39;G&#39;: s += &#39;0010&#39;
                        elif nt == &#39;T&#39;: s += &#39;0001&#39;
                        else: raise()
                s += str(label)
                data.append(s)
        return data</code></pre>
</details>
</dd>
<dt id="seqio.random_dna"><code class="name flex">
<span>def <span class="ident">random_dna</span></span>(<span>length)</span>
</code></dt>
<dd>
<div class="desc"><p><em>Function that generates random dna sequences based on desired length (not weighted)</em> <br/></p>
<p><em>Returns random dna sequence based on input length (str)</em> <br/></p>
<p><strong>Parameter:</strong></p>
<hr>
<ul>
<li>length &ndash; length of generated dna sequence provided by the user (int)</li>
</ul></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def random_dna(length):
        &#34;&#34;&#34;
        *Function that generates random dna sequences based on desired length (not weighted)* &lt;br/&gt;

        *Returns random dna sequence based on input length (str)* &lt;br/&gt;

        **Parameter:**
        ______________

        + length -- length of generated dna sequence provided by the user (int)

        &#34;&#34;&#34;
        nts = &#39;ACGT&#39;
        seq = []
        for i in range(length):
                seq.append(random.choice(nts))
        return &#39;&#39;.join(seq)</code></pre>
</details>
</dd>
<dt id="seqio.read_fasta"><code class="name flex">
<span>def <span class="ident">read_fasta</span></span>(<span>filename)</span>
</code></dt>
<dd>
<div class="desc"><p><em>Function that reads fasta files</em> <br/></p>
<p><em>Returns sequence name followed by sequences from the imported file</em> <br/></p>
<p><strong>Parameters:</strong></p>
<hr>
<ul>
<li>filename &ndash; path to the file (str)</li>
</ul></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def read_fasta(filename):
        &#34;&#34;&#34;
        *Function that reads fasta files* &lt;br/&gt;

        *Returns sequence name followed by sequences from the imported file* &lt;br/&gt;

        **Parameters:**
        _______________

        + filename -- path to the file (str)
        &#34;&#34;&#34;

        name = None
        seqs = []

        fp = None
        if filename == &#39;-&#39;:
                fp = sys.stdin
        elif filename.endswith(&#39;.gz&#39;):
                fp = gzip.open(filename, &#39;rt&#39;)
        else:
                fp = open(filename)

        for line in fp.readlines():
                line = line.rstrip()
                if line.startswith(&#39;&gt;&#39;):
                        if len(seqs) &gt; 0:
                                seq = &#39;&#39;.join(seqs)
                                yield(name, seq)
                                name = line[1:]
                                seqs = []
                        else:
                                name = line[1:]
                else:
                        seqs.append(line)
        yield(name, &#39;&#39;.join(seqs))
        fp.close()</code></pre>
</details>
</dd>
<dt id="seqio.read_raw"><code class="name flex">
<span>def <span class="ident">read_raw</span></span>(<span>filename)</span>
</code></dt>
<dd>
<div class="desc"><p><em>Function that returns sequencing data in fasta format</em> <br/></p>
<p><em>Returns unfiltered names and sequences</em> <br/></p>
<p><strong>Parameter:</strong></p>
<hr>
<ul>
<li>filename &ndash; path to the fasta file containing sequences (str)</li>
</ul></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def read_raw(filename):
        &#34;&#34;&#34;
        *Function that returns sequencing data in fasta format* &lt;br/&gt;

        *Returns unfiltered names and sequences* &lt;br/&gt;

        **Parameter:**
        ______________

        + filename -- path to the fasta file containing sequences (str)
        &#34;&#34;&#34;

        fp = None
        if filename == &#39;-&#39;:
                fp = sys.stdin
        elif filename.endswith(&#39;.gz&#39;):
                fp = gzip.open(filename, &#39;rt&#39;)
        else:
                fp = open(filename)

        for line in fp.readlines():
                line = line.rstrip()
                yield(line)

        fp.close()</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="seqio.cross_validation" href="#seqio.cross_validation">cross_validation</a></code></li>
<li><code><a title="seqio.fasta2onehot" href="#seqio.fasta2onehot">fasta2onehot</a></code></li>
<li><code><a title="seqio.random_dna" href="#seqio.random_dna">random_dna</a></code></li>
<li><code><a title="seqio.read_fasta" href="#seqio.read_fasta">read_fasta</a></code></li>
<li><code><a title="seqio.read_raw" href="#seqio.read_raw">read_raw</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>